### PR TITLE
NPE in ASTRewriteFlattener as return value of GuardedPattern.getPattern() is null 

### PR DESCRIPTION
**Pull request by subyssurendran666**
Issues: #1483 

## What it does

Fixes: #1483 

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/GuardedPattern.java [hunk 1]
```diff
@@ -199,6 +199,16 @@ public static List propertyDescriptors(int apiLevel, boolean previewEnabled) {
 	 */
 	public Expression getExpression() {
 		supportedOnlyIn21();
+		if(this.conditonalExpression == null) {
+            //lazy init must be thread-safe for readers
+            synchronized (this) {
+                if( this.conditonalExpression == null) {
+                    preLazyInit();
+                    this.conditonalExpression = this.ast.newInfixExpression();
```
Comment (mpalat) [inline org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/GuardedPattern.java:207]: Please use null literal
                    this.conditonalExpression = this.ast.newNullLiteral();

File: org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/rewrite/describing/ASTRewritingRecordPatternTest.java [hunk 2]
```diff
@@ -585,4 +585,63 @@ public void testModifyRecordInstanceOfPattern() throws Exception {
 		assertEqualString(preview, buf.toString());
 	}
 
+	public void testNPEinASTRewriteFlattener() throws Exception {
```
Comment (mpalat) [inline org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/rewrite/describing/ASTRewritingRecordPatternTest.java:588]: ASTRewritingSwitchPatternTest is the right place for this test as there is no RecordPattern in the test case. Please move this test there

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/GuardedPattern.java [hunk 3]
```diff
@@ -213,6 +223,16 @@ public Expression getExpression() {
 	 */
 	public Pattern getPattern() {
 		supportedOnlyIn21();
+		if(this.pattern == null) {
+            // lazy init must be thread-safe for readers
+            synchronized (this) {
+                if(this.pattern == null) {
+                    preLazyInit();
+                    this.pattern = this.ast.newTypePattern();
```
Comment (mpalat) [inline org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/GuardedPattern.java:231]: Please use:
                    this.pattern = this.ast.newNullPattern();

**Review by mpalat — CHANGES_REQUESTED** [review]

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/GuardedPattern.java [hunk 4]
```diff
@@ -199,6 +199,16 @@ public static List propertyDescriptors(int apiLevel, boolean previewEnabled) {
 	 */
 	public Expression getExpression() {
 		supportedOnlyIn21();
+		if(this.conditonalExpression == null) {
+            //lazy init must be thread-safe for readers
+            synchronized (this) {
```
[the remark this answers is not in the document]
Comment (subyssurendran666) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/GuardedPattern.java]: Corrected the indentation.

File: org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/rewrite/describing/ASTRewritingRecordPatternTest.java (hunk 2 above)
Comment (subyssurendran666) [reply org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/rewrite/describing/ASTRewritingRecordPatternTest.java]: the test case moved to **ASTRewritingSwitchPatternTest**.

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/GuardedPattern.java (hunk 3 above)
Comment (subyssurendran666) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/GuardedPattern.java]: `newTypePattern()` has been modified to `newNullPattern()`.

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/GuardedPattern.java [hunk 5]
```diff
@@ -199,6 +199,16 @@ public static List propertyDescriptors(int apiLevel, boolean previewEnabled) {
...
 	public Expression getExpression() {
 		supportedOnlyIn21();
+		if(this.conditonalExpression == null) {
+			//lazy init must be thread-safe for readers
+			synchronized (this) {
+				if(this.conditonalExpression == null) {
+					preLazyInit();
+					this.conditonalExpression = this.ast.newNullLiteral();
+					postLazyInit(this.pattern, EXPRESSION_PROPERTY);
+                }
```
Comment (mpalat) [inline org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/GuardedPattern.java:209]: please correct the indentation

**Review by mpalat — CHANGES_REQUESTED** [review]
Thanks for incorporating the changes - please find a follow-up minor comment.
Also, use the original code in the issue and see whether any more tests needed.
Thanks

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/GuardedPattern.java (hunk 5 above)
Comment (subyssurendran666) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/GuardedPattern.java]: Done

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/GuardedPattern.java [hunk 6]
```diff
@@ -213,6 +223,16 @@ public Expression getExpression() {
 	 */
 	public Pattern getPattern() {
 		supportedOnlyIn21();
+		if(this.pattern == null) {
```
Comment (mpalat) [inline org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/GuardedPattern.java:226]: please put a space after if

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/GuardedPattern.java [hunk 7]
```diff
@@ -199,6 +199,16 @@ public static List propertyDescriptors(int apiLevel, boolean previewEnabled) {
 	 */
 	public Expression getExpression() {
 		supportedOnlyIn21();
+		if(this.conditonalExpression == null) {
+			//lazy init must be thread-safe for readers
+			synchronized (this) {
+				if(this.conditonalExpression == null) {
```
Comment (mpalat) [inline org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/GuardedPattern.java:205]: space after if

**Review by mpalat — CHANGES_REQUESTED** [review]

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/GuardedPattern.java (hunk 6 above)
Comment (subyssurendran666) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/GuardedPattern.java]: Done

**Review by mpalat — APPROVED** [review]
Looks good. +1 to merge once the tests pass

**Pull request merged**